sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -185,8 +183,21 @@ case class DataSource(
     }
 
     SchemaUtils.checkColumnNameDuplication(
-      (dataSchema ++ partitionSchema).map(_.name), "in the data schema and the partition schema",
-      sparkSession.sessionState.conf.caseSensitiveAnalysis)
+      dataSchema.map(_.name), "in the data schema", equality)
+    SchemaUtils.checkColumnNameDuplication(
+      partitionSchema.map(_.name), "in the partition schema", equality)
+
+    // We just print a waring message if the data schema and partition schema have the duplicate
+    // columns. This is because we allow users to do so in the previous Spark releases and
+    // we have the existing tests for the cases (e.g., `ParquetHadoopFsRelationSuite`).
+    // See SPARK-18108 and SPARK-21144 for related discussions.
+    try {
+      SchemaUtils.checkColumnNameDuplication(
```

Comment:
Also, I think this check (https://github.com/apache/spark/pull/17758/files#diff-f9858c2d9d1a3c3e48753ef675bc865aR66) could cover write cases.